### DISCUSSION TITLE
Add Find and Replace Feature

### DISCUSSION THREAD

**Discussion by valpan1** [cssnr/link-extractor#101] [Feature Requests]
hi.

windows 11. firefox. all latest.

when in "Lazy Load" open links extracted from https://www.tradingview.com/crypto-screener/ with delay of 1 second 
every page change color from black to white which turns into stroboscopic like effect and destroy attention to my work.
please let it be all white.
(in tab rotating extension i can't set big delay coz pages are heavily scripted and must be fully loaded what i check them)

next i'll tell you what i do and you will better understand what i need.

i grab links from https://www.tradingview.com/crypto-screener/
and the link is like this https://www.tradingview.com/symbols/BINANCE-REEFUSDT.P/ 
but i need the link to be like this https://www.tradingview.com/chart/B3X7TA06/?symbol=BINANCE%3AREEFUSDT.P 

so i mass modify links in text editor changing all https://www.tradingview.com/symbols/BINANCE- to https://www.tradingview.com/chart/B3X7TA06/?symbol=BINANCE%3A 
and then changing USDT.P/ to USDT.P 

then i open all modified links. 

so every time (i check coins often) i do a lot of annoying identical work which makes me lazy and reduces my productivity.
i wish i push 1 button and all links are grabbed and modified by presets and opened.

thank you!

**Comment by smashedr** [comment]
I see 2 requests/issues here:

1. Lazy Loaded tabs are displaying a black background, color `#1c1b21`, by default, which is causing black screen then white screens as the page loads with a white background.

@valpan1 Can you confirm that you are on using a `Light` system theme and all the web extension pages are white?

The best solution I have here, is to check the system theme, and if it is set to Light, use a white background on lazy loaded tabs, or if it is Dark, use a black background. This should make a white -> white transition for Light modes, and a black -> black transition for Dark modes.

2. You want a way to transform links from one format to another, `https://www.tradingview.com/symbols/BINANCE-REEFUSDT.P/` -> `https://www.tradingview.com/chart/B3X7TA06/?symbol=BINANCE%3AREEFUSDT.P`

I will need to do some brainstorming on this to determine a way to implement this feature easily. It can probably be done with regular expression match groups.

---

I can easily implement a fix for the first issue in the next version, however, it may take some time to develop a working solution for the 2nd request. I do see many uses for the 2nd request, so will continue to work on a solution for it.

Thank you very much for the feedback.

**Comment by smashedr** [comment]
For the 1st request, I have merged a fix and will cut a release ASAP:  #102

For the 2nd request, I have a crude implementation, but I assume this is what you are looking  for:

![firefox-20240913-180651828](https://github.com/user-attachments/assets/fbeb1e19-c137-4a18-8038-fb2de0a07292)

You can track the progress of the 2nd request: #103

**Comment by valpan1** [comment]
1. windows 11 theme is light. all pages of Link Extractor are white. lazy load pages are black. 
2. on your .gif picture links modified to https://www.tradingview.com/chart/B3X7TA06/?symbol=BINANCE-REEFUSDT.P but no such coin. 
right one is https://www.tradingview.com/chart/B3X7TA06/?symbol=BINANCE%3AREEFUSDT.P 
as you see not "-" but "%3A". 

(sorry but result link is not usable for you anyway coz B3X7TA06 is only for me. but i think all clear here).

also, please, make data for replacing stored. so i won't need to copy/paste regular expressions every time. 

thank you!

**Comment by valpan1** [comment]
i think, at least for simplified configuring of regular expressions, we need to have some lines for replacements (find, replace, execute). 
sometimes we need to replace not one but some parts of the link. 
making it in 1 regular expression can be super tricky or even impossible without step by step replacement. 

in my example we need to delete "/" at the end too and looks like it is done with one regular expression. 
but what if the task is more complicated...

**Comment by valpan1** [comment]
please, release version just for this current replacement. so many people can use it asap. 
multiple regular expressions for step by step replace you can add later. 
thank you!

p.s you can even promote link-extractor as "good for tradingview screener bulk actions". you know millions of people every day trying to earn with tradingview. only search by "tradingview" keyword will give tons of users for link-extractor . 

or you can go even further and make standalone extension for tradingview screener with "tradingview" keyword in the title and example how to use it in the description (coz not everyone knows this method is possible and useful). this will rock!

**Comment by valpan1** [comment]
heh) here is new interesting thoughts. 

you can see we have access to chart from the screener. but the problem is this is not "usual" link and link-extractor does not see this. but you can click it. 

to be clear, all we do about regular expressions is just because of this. if in the screener we have detectable by link-extractor direct links to charts then no regular expressions need and we just parse and open them. 
this idea you can put to the base of standalone extension for tradingview screener. 
**"Tradingview screener bulk charts opener" - super simple, everyone want :D**

![rjhguvybertytvidrutuyhdbunthvxigurertrrty456](https://github.com/user-attachments/assets/ef09d585-eef5-4f9f-b92c-5b137043d7ad)

**Comment by valpan1** [comment]
also, from my own experience, we need to open links in the order shown in the screener. not alphabetical order. 
this is because sorting coins by different parameters matters and when we open 100 coins to check charts we always want to check the most suitable first and not waste time in random checks.

**Comment by valpan1** [comment]
to make "super combine all in one" you can add feature to open links one by one with custom delay in seconds. same as i do now with tab rotator extension. it clicks your "lazy load" tabs one by one and i get many pages fully loaded and ready for check. 
otherwise, by doing this by hands i need to click a lot on tabs and wait when all indicators on chart are fully loaded (sometimes 15 seconds per page).

one by one coz open all pages together will overload system memory and cpu. for the same reason we need to limit amount of pages open in one task (one "open" button push), or have "stop" button to stop opening when we want to continue it later from any tab to the right. 
open and stop button is the same one.

**Comment by smashedr** [comment]
I did not have time to test and cut a release today, but will get one out tomorrow. This will allow you to use the feature, and provide better feedback.

**Comment by smashedr** [comment]
I have published the current implementation. **It is live on Firefox now**.
Chrome users will need to manually install the addon from the release page, since Google publishing takes a while, by downloading the `.crx` and dragging/dropping it to the `chrome://extensions/` page from here: https://github.com/cssnr/link-extractor/releases/tag/0.7.8

There are a few things to note about the current implementation:

1. There is no feature to save find/replace patterns yet; however, your browser should save them in your form history for now.
2. You can perform multiple find/replaces on the links; however, you have to edit the find/replace inputs each time as there is not yet a way to add additional inputs.

Let me know if you find any bugs or have any issues using the feature. Thanks!

**Comment by valpan1** [comment]
it works)) thank you!

**Comment by valpan1** [comment]
about not alphabetical order i said already. also will be good to have open in random order.

**Comment by valpan1** [comment]
when open lazy tabs "favicon" of websites shown. which confusing me coz i don't know which tab is lazy and which is already with loaded website. i understood this after changing other extension to your. please change this to no favicon, or as option. (maybe sounds strange but this is how i feel in my work)

we can see which domain lazy tab contains without favicon.

![{1B251630-4A51-4DC5-A949-769BC66302F9}](https://github.com/user-attachments/assets/61ad673f-6037-4ac8-bf31-3ba12334b345)

**Comment by smashedr** [comment]
- Alphabetical Order Issue

You can click on the `Links` heading to toggle between 3 states; alphabetical order, reverse alphabetical order, and order the links were extracted on the page.
There is a screen shot in this issue: https://github.com/cssnr/link-extractor/issues/98
As well as information on the FAQ (you may need to refresh the page after loading due to a previous SW bug): https://link-extractor.cssnr.com/faq/#orderLinks

- Favicon on Lazy Loaded Tabs

Personally I like the favicon, but I can see how this can be opinionated. I can add an option in the next release, that lets you enable or disable showing the favicon on lazy loaded tabs.
